File: programmers/lv2/카펫.py
"""
    lv.2 카펫
    
    1.  가로 x 세로 로 성립 가능한 모든 조합에 대해 탐색.
    2.  해당 조합에 대해, 외곽을 잘라낸 넓이가 yellow와 동일하면 탐색 종료.
"""


def solution(brown, yellow):
    answer = []

    total = brown + yellow

    for col in range(3, int(total**0.5) + 1):
        if total % col != 0:
            continue

        row = total // col

        if (row - 2) * (col - 2) == yellow:
            answer = [row, col]
            break

    return answer

File: programmers/lv2/카펫.js
/* 
lv.2 카펫

1.  테두리/내부의 타일 개수와 총 넓이가 정해져 있으면 가능한 가로/세로는 정확히 하나의 케이스만 존재한다.
2.  총 넓이에서 테두리를 뺀 내부 넓이를 가능한 모든 가로x세로의 경우를 따져봐도 시간에 초과하지 않는다.
 */

function solution(brown, yellow) {
  var answer = [];

  const total = brown + yellow;

  for (let i = 3; i < total / 3 + 1; i++) {
    if (!Number.isInteger(total / i)) {
      continue;
    }

    if ((i - 2) * (total / i - 2) === yellow) {
      answer = [total / i, i];
      break;
    }
  }

  return answer;
}

const brown = 10;
const yellow = 2;

console.log(solution(brown, yellow));
